# Exercise 1: Business cycle correlations

For this exercise, you'll be using macroeconomic data from the folder `data/FRED`.

1.  There are seven decade-specific files named `FRED_monthly_19X0.csv` where `X` identifies the decade (`X` takes on the values 5, 6, 7, 8, 9, 0, 1). Write a loop that reads in all seven files as DataFrames and store them in a list.

    *Hint:* Recall from the lecture that you should use `pd.read_csv(..., parse_dates=['DATE'])` to automatically parse strings stored in the `DATE` column as dates.
2.  Use [`pd.concat()`](https://pandas.pydata.org/docs/reference/api/pandas.concat.html) to concate these data sets into a single `DataFrame` and set the `DATE` column as the index.
3.  You realize that your data does not include GDP since this variable is only reported at quarterly frequency.
    Load the GDP data from the file `GDP.csv` and merge it with your monthly data using an _inner join_.
4.  You want to compute how (percent) changes of the variables in your data correlate with percent changes in GDP.

    1. Create a _new_ `DataFrame` which contains the percent changes in CPI and GDP (using 
    [`pct_change()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pct_change.html),
    and the absolute changes for the remaining variables (using 
    [`diff()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.diff.html)).
    2.  Compute the correlation of the percent changes in GDP with the (percent) changes of all other variables using [`corr()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.corr.html). What does the sign and magnitude of the correlation coefficient tell you?



In [10]:
#Part 1
import pandas as pd

data=[]
DATA_PATH = '../../data/FRED'
for X in range(5,10):
    filename=f'{DATA_PATH}/FRED_monthly_19{X}0.csv'
    df= pd.read_csv(filename,parse_dates=['DATE'])
    data.append(df)
for X in range(0,2):
    filename=f'{DATA_PATH}/FRED_monthly_20{X}0.csv'
    df= pd.read_csv(filename,parse_dates=['DATE'])
    data.append(df)

In [8]:
len(data)

7

In [11]:
#Part 2=concatenation into a single Data frame
df=pd.concat(data,ignore_index=True)

In [12]:
#setting index as date
df=df.set_index('DATE')


In [13]:
df.loc['1950']

CPI  UNRATE  FEDFUNDS  REALRATE  LFPART
DATE                                                
1950-01-01  23.5     6.5       NaN       NaN    58.9
1950-02-01  23.6     6.4       NaN       NaN    58.9
1950-03-01  23.6     6.3       NaN       NaN    58.8
1950-04-01  23.6     5.8       NaN       NaN    59.2
1950-05-01  23.8     5.5       NaN       NaN    59.1
1950-06-01  23.9     5.4       NaN       NaN    59.4
1950-07-01  24.1     5.0       NaN       NaN    59.1
1950-08-01  24.2     4.5       NaN       NaN    59.5
1950-09-01  24.3     4.4       NaN       NaN    59.2
1950-10-01  24.5     4.2       NaN       NaN    59.4
1950-11-01  24.6     4.2       NaN       NaN    59.3
1950-12-01  25.0     4.3       NaN       NaN    59.2

In [36]:
#Part 3
filename2=f'{DATA_PATH}/GDP.csv'
df_gdp= pd.read_csv(filename2,parse_dates=['DATE'],index_col='DATE')


In [39]:
#merging the two df
df_merged=pd.merge(df_gdp,df,on='DATE', how='inner')
df_merged

GDP    CPI  UNRATE  FEDFUNDS  REALRATE  LFPART
DATE                                                          
1950-01-01   2346.1   23.5     6.5       NaN       NaN    58.9
1950-04-01   2417.7   23.6     5.8       NaN       NaN    59.2
1950-07-01   2511.1   24.1     5.0       NaN       NaN    59.1
1950-10-01   2559.2   24.5     4.2       NaN       NaN    59.4
1951-01-01   2594.0   25.4     3.7       NaN       NaN    59.1
...             ...    ...     ...       ...       ...     ...
2018-10-01  20304.9  252.8     3.8       2.2      -0.2    62.9
2019-01-01  20431.6  252.6     4.0       2.4       0.6    63.1
2019-04-01  20602.3  255.2     3.7       2.4       3.1    62.8
2019-07-01  20843.3  255.8     3.7       2.4       1.1    63.1
2019-10-01  20985.4  257.2     3.6       1.8      -0.0    63.3

[280 rows x 6 columns]

In [26]:
df

CPI  UNRATE  FEDFUNDS  REALRATE  LFPART
DATE                                                 
1950-01-01   23.5     6.5       NaN       NaN    58.9
1950-02-01   23.6     6.4       NaN       NaN    58.9
1950-03-01   23.6     6.3       NaN       NaN    58.8
1950-04-01   23.6     5.8       NaN       NaN    59.2
1950-05-01   23.8     5.5       NaN       NaN    59.1
...           ...     ...       ...       ...     ...
2019-08-01  256.0     3.6       2.1       0.6    63.1
2019-09-01  256.4     3.5       2.0       0.3    63.2
2019-10-01  257.2     3.6       1.8      -0.0    63.3
2019-11-01  257.9     3.6       1.6      -0.2    63.3
2019-12-01  258.6     3.6       1.6      -0.3    63.3

[840 rows x 5 columns]

In [ ]:
#Part 4 percentage changes
df_pct_gdp= df_merged['GDP'].pct_change()
df_pct_cpi=df_merged['CPI'].pct_change()

df_pct_gdp

DATE
1950-01-01         NaN
1950-04-01    0.030519
1950-07-01    0.038632
1950-10-01    0.019155
1951-01-01    0.013598
                ...   
2018-10-01    0.001415
2019-01-01    0.006240
2019-04-01    0.008355
2019-07-01    0.011698
2019-10-01    0.006818
Name: GDP, Length: 280, dtype: float64

In [41]:
df_pct_cpi

DATE
1950-01-01         NaN
1950-04-01    0.004255
1950-07-01    0.021186
1950-10-01    0.016598
1951-01-01    0.036735
                ...   
2018-10-01    0.006369
2019-01-01   -0.000791
2019-04-01    0.010293
2019-07-01    0.002351
2019-10-01    0.005473
Name: CPI, Length: 280, dtype: float64

In [44]:
#percent change
df_changes= df_merged[['CPI','GDP']].pct_change()*100.0
df_changes


CPI       GDP
DATE                          
1950-01-01       NaN       NaN
1950-04-01  0.425532  3.051873
1950-07-01  2.118644  3.863176
1950-10-01  1.659751  1.915495
1951-01-01  3.673469  1.359800
...              ...       ...
2018-10-01  0.636943  0.141545
2019-01-01 -0.079114  0.623987
2019-04-01  1.029295  0.835471
2019-07-01  0.235110  1.169772
2019-10-01  0.547303  0.681754

[280 rows x 2 columns]

In [46]:
#absolute changes 
variables= ['UNRATE','FEDFUNDS','REALRATE','LFPART']
df_changes[variables]=df_merged[variables].diff()
df_changes


CPI       GDP  UNRATE  FEDFUNDS  REALRATE  LFPART
DATE                                                              
1950-01-01       NaN       NaN     NaN       NaN       NaN     NaN
1950-04-01  0.425532  3.051873    -0.7       NaN       NaN     0.3
1950-07-01  2.118644  3.863176    -0.8       NaN       NaN    -0.1
1950-10-01  1.659751  1.915495    -0.8       NaN       NaN     0.3
1951-01-01  3.673469  1.359800    -0.5       NaN       NaN    -0.3
...              ...       ...     ...       ...       ...     ...
2018-10-01  0.636943  0.141545     0.0       0.3      -0.1    -0.1
2019-01-01 -0.079114  0.623987     0.2       0.2       0.8     0.2
2019-04-01  1.029295  0.835471    -0.3       0.0       2.5    -0.3
2019-07-01  0.235110  1.169772     0.0       0.0      -2.0     0.3
2019-10-01  0.547303  0.681754    -0.1      -0.6      -1.1     0.2

[280 rows x 6 columns]

In [47]:
df_changes.corr()

CPI       GDP    UNRATE  FEDFUNDS  REALRATE    LFPART
CPI       1.000000 -0.113091  0.054552  0.124710 -0.254783  0.101229
GDP      -0.113091  1.000000 -0.564872  0.206370  0.074500  0.019639
UNRATE    0.054552 -0.564872  1.000000 -0.348058 -0.090960  0.087339
FEDFUNDS  0.124710  0.206370 -0.348058  1.000000  0.428688  0.042001
REALRATE -0.254783  0.074500 -0.090960  0.428688  1.000000  0.024850
LFPART    0.101229  0.019639  0.087339  0.042001  0.024850  1.000000

In [48]:
df_changes.corr().loc['GDP']

CPI        -0.113091
GDP         1.000000
UNRATE     -0.564872
FEDFUNDS    0.206370
REALRATE    0.074500
LFPART      0.019639
Name: GDP, dtype: float64

***
# Exercise 2: Loading many data files

In the previous exercise, you loaded the individual files by specifing an explicit list of file names. This can become tedious or infeasible if your data is spread across many files with varying file name patterns. Python offers the possibility to iterate over all files in a directory (for example, using [`os.listdir()`](https://docs.python.org/3/library/os.html#os.listdir)),
or to iterate over files that match a pattern, for example using [`glob.glob()`](https://docs.python.org/3/library/glob.html).

Repeat parts (1) and (2) from the previous exercise, but now iterate over the input files using 
[`glob.glob()`](https://docs.python.org/3/library/glob.html). You'll need to use a wildcard `*` and make sure to match only the relevant files in `data/FRED`, i.e., those that start with `FRED_monthly_1` or `FRED_monthly_2`.

In [54]:
pattern=f'{DATA_PATH}/FRED_monthly_[12]*.csv'

import glob 

data= []
for file in glob.glob(pattern):
    print(f'processing file {file}')
    df= pd.read_csv(file)
    data.append(df)


processing file ../../data/FRED\FRED_monthly_1950.csv
processing file ../../data/FRED\FRED_monthly_1960.csv
processing file ../../data/FRED\FRED_monthly_1970.csv
processing file ../../data/FRED\FRED_monthly_1980.csv
processing file ../../data/FRED\FRED_monthly_1990.csv
processing file ../../data/FRED\FRED_monthly_2000.csv
processing file ../../data/FRED\FRED_monthly_2010.csv


In [55]:
#merge this to a single dataframe
df=pd.concat(data, ignore_index=True)
df

DATE    CPI  UNRATE  FEDFUNDS  REALRATE  LFPART
0    1950-01-01   23.5     6.5       NaN       NaN    58.9
1    1950-02-01   23.6     6.4       NaN       NaN    58.9
2    1950-03-01   23.6     6.3       NaN       NaN    58.8
3    1950-04-01   23.6     5.8       NaN       NaN    59.2
4    1950-05-01   23.8     5.5       NaN       NaN    59.1
..          ...    ...     ...       ...       ...     ...
835  2019-08-01  256.0     3.6       2.1       0.6    63.1
836  2019-09-01  256.4     3.5       2.0       0.3    63.2
837  2019-10-01  257.2     3.6       1.8      -0.0    63.3
838  2019-11-01  257.9     3.6       1.6      -0.2    63.3
839  2019-12-01  258.6     3.6       1.6      -0.3    63.3

[840 rows x 6 columns]

In [56]:
df.sort_values('DATE').reset_index(drop=True)

DATE    CPI  UNRATE  FEDFUNDS  REALRATE  LFPART
0    1950-01-01   23.5     6.5       NaN       NaN    58.9
1    1950-02-01   23.6     6.4       NaN       NaN    58.9
2    1950-03-01   23.6     6.3       NaN       NaN    58.8
3    1950-04-01   23.6     5.8       NaN       NaN    59.2
4    1950-05-01   23.8     5.5       NaN       NaN    59.1
..          ...    ...     ...       ...       ...     ...
835  2019-08-01  256.0     3.6       2.1       0.6    63.1
836  2019-09-01  256.4     3.5       2.0       0.3    63.2
837  2019-10-01  257.2     3.6       1.8      -0.0    63.3
838  2019-11-01  257.9     3.6       1.6      -0.2    63.3
839  2019-12-01  258.6     3.6       1.6      -0.3    63.3

[840 rows x 6 columns]

***
# Exercise 3: Weekly returns of the magnificent seven

In this exercise, you are asked to analyze the weekly stockmarket returns
of the so-called magnificent 7 which are some of the most successful tech companies 
of the last decades years:
Apple (AAPL), Amazon (AMZN), Alphabet/Google (GOOGL), Meta (META), Microsoft (MSFT), Nvidia (NVDA), and Tesla (TSLA).

The data for this exercise is located in the folder `data/stockmarket/`.

1.  For each of the seven stocks listed above, there is a corresponding 
    CSV file in this directory (based on the ticker symbol).

    1.  For each ticker symbol, load the corresponding CSV file and make sure 
        that the `Date` is set as the index.

        The DataFrame has two columns, `Open` and `Close`, which contain the 
        opening and closing price for each trading day.

    3.  Use [`resample()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.resample.html)
        to resample the daily data to a weekly frequency by specifying `resample('W')`,
        and compute the weekly returns in percent:

        $$
        \text{Weekly returns} = \frac{\text{Close price on last day} - \text{Open price on first day}}{\text{Open price on first day}} \times 100
        $$

        *Hint:* You can obtain the first and last observation using the 
        [`first()`](https://pandas.pydata.org/docs/reference/api/pandas.core.resample.Resampler.first.html) and 
        [`last()`](https://pandas.pydata.org/docs/reference/api/pandas.core.resample.Resampler.last.html)
        methods.

    4.  Append these returns to a list so you can merge them into a single DataFrame later.

2.  Merge the list of weekly returns you computed into a single DataFrame.
    Keep only the intersection of dates available for all 7 stocks.

    *Hint:* This can be achieved using either 
    [`pd.concat()`](https://pandas.pydata.org/docs/reference/api/pandas.concat.html),
    [`pd.merge()`](https://pandas.pydata.org/docs/reference/api/pandas.merge.html), or 
    [`DataFrame.join()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.join.html).

3.  Finally, you are interested in how the weekly returns are correlated across 
    the 7 stocks. 

    1.  Compute and report the pairwise correlations using 
        [DataFrame.corr()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.corr.html).

    2.  Create a figure with 7-by-7 subplots showing the pairwise scatter plots of weekly returns 
        for each combination of stocks.

        You can do this either with the
        [`scatter_matrix()`](https://pandas.pydata.org/docs/reference/api/pandas.plotting.scatter_matrix.html) function contained in `pandas.plotting`, 
        or manually build the figure using Matplotlib functions.

    3.  **[Advanced]**
        In each of the subplots, add a text that reports the pairwise correlation
        for these stocks which you computed earlier.
        (e.g., the correlation between returns on AAPL and AMZN is about 0.42,
        so this text should be added to the subplot showing the 
        scatter plot of AAPL vs. AMZN).


In [61]:
DATA_PATH = '../../data/stockmarket'
data=[]
#LOOP OVER TICKER SYMBOLS AND PERFORM TASKS IN PART 1 
tickers=['AAPL','AMZN','GOOGL','META','MSFT','NVDA','TSLA']
for x in tickers:
    filename=f'{DATA_PATH}/{x}.csv'
    print(f'processing file {filename}')
    df=pd.read_csv(filename,parse_dates=True,index_col='Date')



processing file ../../data/stockmarket/AAPL.csv
processing file ../../data/stockmarket/AMZN.csv
processing file ../../data/stockmarket/GOOGL.csv
processing file ../../data/stockmarket/META.csv
processing file ../../data/stockmarket/MSFT.csv
processing file ../../data/stockmarket/NVDA.csv
processing file ../../data/stockmarket/TSLA.csv


In [ ]:
#part 1, 2 
first= df.resample('W')['Open'].first()
last=df.resample('W')['Close'].last()

returns= (last-first/first)*100


***
# Exercise 4: Decade averages of macro time series


For this exercise, you'll be using macroeconomic data from the folder `data/FRED`.

1.  There are five files containing monthly observations on annual inflation (INFLATION), the Fed Funds rate (FEDFUNDS), the labor force participation rate (LFPART), the 1-year real interest rate (REALRATE) and the unemployment rate (UNRATE).

    1.  Write a loop to import these files and store the individual DataFrames in a list.

        *Hint:* Recall from the lecture that you should use 
        `pd.read_csv(..., parse_dates=['DATE'], index_col='DATE')` to automatically parse strings stored in the `DATE` column as dates and set the `DATE`
        column as the index.

    2.  Use 
        [`pd.concat()`](https://pandas.pydata.org/docs/reference/api/pandas.concat.html)
        to concatenate this list of DataFrames along the column dimension
        using an outer join (`join='outer'`) to obtain a merged data set.

3.  You want to compute the average value of each variable by decade, but you want to include only decades without _any_ missing values for _all_ variables.

    1.  Create a variable `Decade` which stores the decade (1940, 1950, ...) for each observation.

        *Hint:* You should have set the `DATE` as the `DataFrame` index. Then you can access the calendar year using the attribute `df.index.year` which can be used to compute the decade.

    2.  Create an indicator variable which takes on the value `True` 
        whenever all observations (all columns) for a given date are non-missing, and `False`
        if at least one variable has a missing observation. 

    3.  Aggregate this indicator to decades using a
    [`groupby()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html) so that the indicator takes on the value `True` whenever
    _all_ variables in a given decade have no missing values, and `False`
    otherwise.

        *Hint:* You can use the 
        [`all()`](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.all.html) aggregation for this.

    4.  Merge this decade-level indicator data back into the original `DataFrame` (_many-to-one_ merge). 
4.  Using this indicator, drop all observations which are in a decade with missing values.
5.  Compute the decade average for each variable.

**Challenge**

-   Your pandas guru friend claims that all the steps in 2.2 to 2.4 can be done with a single one-liner using [`transform()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.transform.html). Can you come up with a solution?

    

***
# Exercise 5: Merging additional Titanic data

In this exercise, you'll be working with the the original Titanic data set in `titanic.csv` and additional (partly fictitious) information on passengers stored in `titanic-additional.csv`, both located in the `data/` folder.

The goal of the exercise is to calculate the survival rates by country of residence (for this exercise we restrict ourselves to the UK, so these will be England, Scotland, etc.).

1.  Load the `titanic.csv` and `titanic-additional.csv` into two DataFrames.

    Inspect the columns contained in both data sets. As you can see, the original data contains the full name including the title
    and potentially maiden name (for married women) in a single column.
    The additional data contains this information in separate columns.
    You want to merge these data sets, but you first need to create common keys in both DataFrames.

2.  Since the only common information is the name, you'll need to extract the individual name components from the original DataFrame
    and use these as merge keys.

    Focusing only on men (who have names that are much easier to parse), split the `Name` column into the tokens 
    `Title`, `FirstName` and `LastName`, just like the columns in the second DataFrame.

    *Hint:* This is the same task as in the last exercise in Workshop 2. You can just use your solution here.

3.  Merge the two data sets based on the columns `Title`, `FirstName` and `LastName` you just created using a _left join_ (_one-to-one_ merge).
    Tabulate the columns and the number of non-missing observations to make sure that merging worked. 

    *Note:* The additional data set contains address information only for passengers from the UK, so some of these fields will be missing.

4.  You are now in a position to merge the country of residence (_many-to-one_ merge). Load the country data from `UK_post_codes.csv` which contains 
    the UK post code prefix (which you can ignore), the corresponding city, and the corresponding country.

    Merge this data with your passenger data set using a _left join_ (what is the correct merge key?).

5.  Tabulate the number of observations by `Country`, including the number of observations with missing `Country` (these are passengers residing outside the UK).

    Finally, compute the mean survival rate by country.